sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs

```diff
@@ -976,13 +976,51 @@ private async Task<AmqpSession> CreateAndOpenSessionAsync(AmqpConnection connect
         ///
         /// <param name="target">The target AMQP object to open.</param>
         /// <param name="timeout">The timeout to apply when opening the link.</param>
+        /// <param name="cancellationToken">Token to signal cancellation of the operation.</param>
         protected virtual async Task OpenAmqpObjectAsync(
             AmqpObject target,
-            TimeSpan timeout)
+            TimeSpan timeout,
+            CancellationToken cancellationToken)
         {
             try
             {
-                await target.OpenAsync(timeout).ConfigureAwait(false);
+                var openObjectCompletionSource = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                using var registration = cancellationToken.Register(static state =>
+                {
+                    var tcs = (TaskCompletionSource<object>)state;
+                    tcs.TrySetCanceled();
+                }, openObjectCompletionSource, useSynchronizationContext: false);
+
+                _ = Task.Factory
+                    .FromAsync<(AmqpObject, TimeSpan, TaskCompletionSource<object>)>
+                    (
+                        static (arguments, callback, state) =>
```

Comment:
```suggestion
                        static (arguments, callback, _) =>
```

This could by the way also be applied in my code

Reply:
I ended up just using OpenAsync instead of the Open/Close methods.

Reply:
No, I just found it harder to read and we were already using the OpenAsync method before. I'm not sure if there is an equivalent for Receive though.

Reply:
Unfortunately there is none

Reply:
I could contribute the method to the lib if you want

Reply:
I don't think it is a big deal really - especially since the code is already written and tested 😄

Reply:
Actually, there is an async style API - https://github.com/Azure/azure-amqp/blob/hotfix/Microsoft.Azure.Amqp/Amqp/ReceivingAmqpLink.cs#L121

Reply:
I decompiled the current version that we used here and couldn't find that. I did not look into the latest updates. Nice find